### DISCUSSION TITLE
Backup on S3 not working

### DISCUSSION THREAD

**Discussion by cbarria** [cloudnative-pg/cloudnative-pg#1175] [General]
Hi everyone, been trying to deploy cloudnativepg for a while now and suddenly when i was almost ready, my backups started failing... and i haven't been able to fix them yet, i believe there is an issue with the fork/exec command when calling the archiver from barman, here are the logs from the pod when trying to backup.
`{
  "level": "error",
  "ts": 1663093503.9465349,
  "logger": "wal-archive",
  "msg": "Error invoking barman-cloud-check-wal-archive",
  "logging_pod": "cnpg-cluster-1",
  "walName": "000000010000000000000001",
  "currentPrimary": "cnpg-cluster-1",
  "targetPrimary": "cnpg-cluster-1",
  "options": [
    "--cloud-provider",
    "aws-s3",
    "s3://cnpg-backups/",
    "cnpg-cluster"
  ],
  "exitCode": -1,
  "error": "fork/exec /usr/local/bin/barman-cloud-check-wal-archive: invalid argument",
  "stacktrace": "github.com/cloudnative-pg/cloudnative-pg/pkg/management/barman/archiver.(*WALArchiver).CheckWalArchive\n\tpkg/management/barman/archiver/archiver.go:227\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.run\n\tinternal/cmd/manager/walarchive/cmd.go:176\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:79\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902\nmain.main\n\tcmd/manager/main.go:61\nruntime.main\n\t/opt/hostedtoolcache/go/1.17.8/x64/src/runtime/proc.go:255"
}`

If i enter this command manually inside the pod when the same ENV variables set, the command executes fine and returns 0

`postgres@cnpg-cluster-1:/$ barman-cloud-check-wal-archive --cloud-provider aws-s3 s3://cnpg-backups/ cnpg-cluster
postgres@cnpg-cluster-1:/$ echo $?
0`

the "invalid argument" i believe it comes from the fork/exec part of it, our environment is pretty standard setup.

credentials and bucket config were tested manually, they just fail when are executed by the pod.

these are the following logs that fail...

`{
  "level": "error",
  "ts": 1663094877.9874961,
  "msg": "while barman-cloud-check-wal-archive",
  "logging_pod": "cnpg-cluster-1",
  "error": "unexpected failure invoking barman-cloud-wal-archive: fork/exec /usr/local/bin/barman-cloud-check-wal-archive: invalid argument",
  "stacktrace": "github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.Error\n\tpkg/management/log/log.go:165\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.run\n\tinternal/cmd/manager/walarchive/cmd.go:177\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:79\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902\nmain.main\n\tcmd/manager/main.go:61\nruntime.main\n\t/opt/hostedtoolcache/go/1.17.8/x64/src/runtime/proc.go:255"
}
{
  "level": "error",
  "ts": 1663094877.9876142,
  "logger": "wal-archive",
  "msg": "failed to run wal-archive command",
  "logging_pod": "cnpg-cluster-1",
  "error": "unexpected failure invoking barman-cloud-wal-archive: fork/exec /usr/local/bin/barman-cloud-check-wal-archive: invalid argument",
  "stacktrace": "github.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:81\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902\nmain.main\n\tcmd/manager/main.go:61\nruntime.main\n\t/opt/hostedtoolcache/go/1.17.8/x64/src/runtime/proc.go:255"
}
{
  "level": "info",
  "ts": 1663094877.9926615,
  "logger": "postgres",
  "msg": "record",
  "logging_pod": "cnpg-cluster-1",
  "record": {
    "log_time": "2022-09-13 18:47:57.992 UTC",
    "process_id": "51",
    "session_id": "6320cad7.33",
    "session_line_num": "107",
    "session_start_time": "2022-09-13 18:24:23 UTC",
    "transaction_id": "0",
    "error_severity": "LOG",
    "sql_state_code": "00000",
    "message": "archive command failed with exit code 1",
    "detail": "The failed archive command was: /controller/manager wal-archive --log-destination /controller/log/postgres.json pg_wal/000000010000000000000001",
    "backend_type": "archiver",
    "query_id": "0"
  }
}
{
  "level": "info",
  "ts": 1663094877.9927022,
  "logger": "postgres",
  "msg": "record",
  "logging_pod": "cnpg-cluster-1",
  "record": {
    "log_time": "2022-09-13 18:47:57.992 UTC",
    "process_id": "51",
    "session_id": "6320cad7.33",
    "session_line_num": "108",
    "session_start_time": "2022-09-13 18:24:23 UTC",
    "transaction_id": "0",
    "error_severity": "WARNING",
    "sql_state_code": "01000",
    "message": "archiving write-ahead log file \"000000010000000000000001\" failed too many times, will try again later",
    "backend_type": "archiver",
    "query_id": "0"
  }
}`

could this possibly be related to something else?

Thanks!

**Comment by shrinedogg** [comment]
Having the same issue on the latest versions.

```
{"level":"error","ts":1663178224.2349117,"logger":"wal-archive","msg":"Error invoking barman-cloud-check-wal-archive","logging_pod":"cloudnative-pg-01-1","currentPrimary":"cloudnative-pg-01-1","targetPrimary":"cloudnative-pg-01-1","options":["--cloud-provider","aws-s3","s3://cnpg/cnpg/","cloudnative-pg-01"],"exitCode":-1,"error":"exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/barman/archiver.(*WALArchiver).CheckWalArchiveDestination\n\tpkg/management/barman/archiver/archiver.go:234\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.checkWalArchive\n\tinternal/cmd/manager/walarchive/cmd.go:379\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.run\n\tinternal/cmd/manager/walarchive/cmd.go:166\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:80\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:872\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918\nmain.main\n\tcmd/manager/main.go:64\nruntime.main\n\t/opt/hostedtoolcache/go/1.18.5/x64/src/runtime/proc.go:250"}


{"level":"error","ts":1663178224.235176,"msg":"while barman-cloud-check-wal-archive","logging_pod":"cloudnative-pg-01-1","error":"unexpected failure invoking barman-cloud-wal-archive: exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.Error\n\tpkg/management/log/log.go:165\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.checkWalArchive\n\tinternal/cmd/manager/walarchive/cmd.go:380\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.run\n\tinternal/cmd/manager/walarchive/cmd.go:166\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:80\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:872\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918\nmain.main\n\tcmd/manager/main.go:64\nruntime.main\n\t/opt/hostedtoolcache/go/1.18.5/x64/src/runtime/proc.go:250"}


{"level":"error","ts":1663178224.2354255,"logger":"wal-archive","msg":"failed to run wal-archive command","logging_pod":"cloudnative-pg-01-1","error":"unexpected failure invoking barman-cloud-wal-archive: exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:82\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:872\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918\nmain.main\n\tcmd/manager/main.go:64\nruntime.main\n\t/opt/hostedtoolcache/go/1.18.5/x64/src/runtime/proc.go:250"}

**Comment by cbarria** [comment]
so sorry its happening to you too... i been trying to deal with this for while, still not root cause from my part.
been trying different versions of helm chart, different images, building custom images but nothing seems to work.
(Kubernetes version v1.23.7)

**Comment by phisco** [comment]
Hi, could one of you please provide some more surrounding logs, grepping for `barman-cloud-wal-archive`, as if the subcommand is starting and logging something either to stdout/stderr it should be logged by the operator. Thanks!

**Comment by cbarria** [comment]
thanks @phisco. i been looking at every logs i can get from the pod and the cluster, could you guide me where else to search? i have my cluster in debug mode.

**Comment by HanssenKai** [comment]
Had a similar issue, where commands were working from within the pod. If you have a error in the pod logs mentioning invalid headers, i.e;
`{"level":"info""logger":"barman-cloud-check-wal-archive","msg":"2022-11-09 07:52:22,423 [85] ERROR: Barman cloud WAL archive check exception: An HTTP Client raised an unhandled exception: Invalid header value b'AWS4-HMAC-SHA256 Credential..."}`  
  
You might have a hidden newline in you credentials secret.

**Comment by sxd** [comment]
This looks more like an issue on the object storage side not supporting encryption, it is this solved already @cbarria @HanssenKai ?

Turning into a discussion

**Comment by ironytr** [comment]
`unexpected failure invoking barman-cloud-wal-archive` i got this error too, then i added `AmazonS3FullAccess` to aws user, now error is gone.
